tasks/Logger.cpp

```diff
@@ -411,9 +388,102 @@ bool Logger::handleExistingFile()
         } while(boost::filesystem::exists(_current_file.get()));
         log(Warning) << "Writing log to " << _current_file.get() << " instead." << endlog();
         return true;
     }
 }
+
+bool Logger::setCurrentFile(::std::string const& value)
+{
+    if(value.empty())
+    {
+      log(Error) << "Could not create log file. Task property _file is empty." << endlog();
+      return false;
+    }
+
+    if (_overwrite_existing_files.get() && _auto_timestamp_files.get())
+    {
+        log(Error) << "The properties overwrite_existing_files and auto_timestamp_files are both set to true, but are mutually exclusive." << endlog();
+        return false;
+    }
+
+    _current_file.set(value);
+
+    if (_auto_timestamp_files.get()) { timestampFile(value); }
+
+    if (boost::filesystem::exists(_current_file.get()))
+    {
+        if (!handleExistingFile(value))
+        {
+            return false;
+        }
+    } else if (_auto_timestamp_files.get())
+    {
+        log(Info) << "Successfully timestamped log file. Writing to: " << _current_file.get() << endlog();
+    }
+    return true;
+}
+
+void Logger::updateLoggers()
+{
+    m_io->close();
+    // The registry has been loaded on construction
+    // Now, create the output file
+    auto_ptr<ofstream> io(new ofstream(_current_file.get().c_str()));
+    auto_ptr<Logfile>  file(new Logfile(*io));
+
+    for (Reports::iterator it = root.begin(); it != root.end(); ++it)
+    {
+        if (!it->logger){
+            it->logger = new Logging::StreamLogger(
+                    it->name, it->type_name, it->cxx_type_name,*(it->registry), it->metadata, *file);
+        }
+        else{
+            delete (it->logger);
+            it->logger = new Logging::StreamLogger(
+                    it->name, it->type_name, it->cxx_type_name,*(it->registry), it->metadata, *file);
+        }
+    }
+
+    m_io   = io.release();
+    m_file = file.release();
+}
+
+bool Logger::setFile(::std::string const &value)
+{
+    std::string prevCurrentFile = _current_file.get();
+
+    try {
+        if (!(_current_file.get().empty()) && (m_io->is_open()))
+        {
+            if (!setCurrentFile(value))
+            {
+                if (_current_file.get() != prevCurrentFile)
+                {
+                    _current_file.set(prevCurrentFile);
+                }
+                return false;
+            }
+            updateLoggers();
+        }
+    }
+    catch(const ofstream::failure& e){
+        log(Error) << "Could not change task property _file: " << e.what() << endlog();
+        if (_current_file.get() != prevCurrentFile)
+        {
+            _current_file.set(prevCurrentFile);
+        }
+        return false;
+    }
+    catch(...){
+        log(Error) << "Could not change task property _file: unidentified error" << endlog();
+        if (_current_file.get() != prevCurrentFile)
+        {
+            _current_file.set(prevCurrentFile);
+        }
+        return false;
+    }
```

Comment:
After a half-failed `updateLoggers()`, the component will be in a bad state, i.e. trying to un-do the change here is just an illusion.

One major issue is that there is too much internal state being changed that needs to be rolled back. Just look at the prevCurrentFile mumbo-jumbo. It's much better to be functional (i.e. not change the object's state) for as long as possible.

- change `currentFile` to return the name of the new current file instead of changing the attribute. Essentially, `computeCurrentFile` would be a const method, and should not be throwing anything (it doesn't so far)
- change `updateLoggers` to take the ofstream object as argument. updateLoggers, internally, should not throw either.

With these two changes, you can be totally functional before `updateLoggers`, which itself should not throw (pseudo-code below):

```
{
    std::string currentFile = computeCurrentFile();
    try {
       io = new ofstream(...);
    }
    catch(const ofstream::failure& e){ 
        /// we haven't changed anything yet, we can safely abort the change and return false
    }

    _current_file.set(currentFile);
    updateLoggers(io);
}
```

Reply:
Should I move updateLoggers (and _current_file.set()) into the try block? VS code marks it as an error, since io is only defined inside try, although theoretically it shouldn't be a problem, because if an error occurs when creating io it will enter the catch block and return false.